pkg/controller/backup_deletion_controller.go

```diff
@@ -0,0 +1,290 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"encoding/json"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/heptio/ark/pkg/apis/ark/v1"
+	"github.com/heptio/ark/pkg/cloudprovider"
+	arkv1client "github.com/heptio/ark/pkg/generated/clientset/versioned/typed/ark/v1"
+	informers "github.com/heptio/ark/pkg/generated/informers/externalversions/ark/v1"
+	listers "github.com/heptio/ark/pkg/generated/listers/ark/v1"
+	"github.com/heptio/ark/pkg/util/kube"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+)
+
+type backupDeletionController struct {
+	*genericController
+
+	deleteBackupRequestClient arkv1client.DeleteBackupRequestsGetter
+	deleteBackupRequestLister listers.DeleteBackupRequestLister
+	backupClient              arkv1client.BackupsGetter
+	snapshotService           cloudprovider.SnapshotService
+	backupService             cloudprovider.BackupService
+	bucket                    string
+	restoreLister             listers.RestoreLister
+	restoreClient             arkv1client.RestoresGetter
+}
+
+// NewBackupDeletionController creates a new backup deletion controller.
+func NewBackupDeletionController(
+	logger logrus.FieldLogger,
+	deleteBackupRequestInformer informers.DeleteBackupRequestInformer,
+	deleteBackupRequestClient arkv1client.DeleteBackupRequestsGetter,
+	backupClient arkv1client.BackupsGetter,
+	snapshotService cloudprovider.SnapshotService,
+	backupService cloudprovider.BackupService,
+	bucket string,
+	restoreInformer informers.RestoreInformer,
+	restoreClient arkv1client.RestoresGetter,
+) Interface {
+	c := &backupDeletionController{
+		genericController:         newGenericController("backup-deletion", logger),
+		deleteBackupRequestClient: deleteBackupRequestClient,
+		deleteBackupRequestLister: deleteBackupRequestInformer.Lister(),
+		backupClient:              backupClient,
+		snapshotService:           snapshotService,
+		backupService:             backupService,
+		bucket:                    bucket,
+		restoreLister:             restoreInformer.Lister(),
+		restoreClient:             restoreClient,
+	}
+
+	c.syncHandler = c.processQueueItem
+	c.cacheSyncWaiters = append(c.cacheSyncWaiters, deleteBackupRequestInformer.Informer().HasSynced, restoreInformer.Informer().HasSynced)
+
+	deleteBackupRequestInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: c.enqueue,
```

Comment:
TODO: need an UpdateFunc (e.g. request is created and modified before the controller sees it, status is still ""/New)